core/src/main/java/org/apache/calcite/rel/mutable/MutableScan.java

```diff
@@ -36,11 +36,12 @@ public static MutableScan of(TableScan scan) {
   @Override public boolean equals(Object obj) {
     return obj == this
         || obj instanceof MutableScan
-        && rel.equals(((MutableScan) obj).rel);
+        && rel.getTable().getQualifiedName().equals(((MutableScan) obj).rel
```

Comment:
Can wet fix the `hashCode` function in `TableScan` directly?

Reply:
Hi @DonnyZone  Thanks a lot for looking into this !
There's a discussion for overriding `equals()&hashCode()` in https://issues.apache.org/jira/browse/CALCITE-3167
The conclusion is to make `equals()&hashCode()` final in `AbstractRelNode ` and  have independent mechanisms to define the equality.

Reply:
@jinxing64 Thanks for clarification. LGTM!